File: __tests__/omarchy-codex-notify.integration.spec.py
import json
import os
from pathlib import Path
import subprocess
import sys
import tempfile
import tomllib
import unittest


ROOT = Path(__file__).resolve().parents[1]


class CodexNotificationsTest(unittest.TestCase):
    def setUp(self):
        self.temp = tempfile.TemporaryDirectory()
        self.addCleanup(self.temp.cleanup)
        self.directory = Path(self.temp.name)
        self.capture = self.directory / "notification.json"
        notifier = self.directory / "omarchy"
        notifier.write_text(
            f"#!{sys.executable}\nimport json, os, sys\n"
            "from pathlib import Path\n"
            "Path(os.environ['NOTIFICATION_CAPTURE']).write_text(json.dumps(sys.argv[1:]))\n"
        )
        notifier.chmod(0o755)
        self.env = dict(os.environ, PATH=f"{self.directory}:{os.environ['PATH']}",
                        CODEX_HOME=str(self.directory), NOTIFICATION_CAPTURE=str(self.capture))
        self.env.pop("TMUX", None)
        self.env.pop("TMUX_PANE", None)
        self.clients = self.directory / "clients.json"
        self.clients.write_text("[]")
        self.actions = self.directory / "actions.jsonl"
        self.env.update(TEST_CLIENTS=str(self.clients), TEST_ACTIONS=str(self.actions))
        hyprctl = self.directory / "hyprctl"
        hyprctl.write_text(
            f"#!{sys.executable}\nimport json, os, sys\nfrom pathlib import Path\n"
            "if sys.argv[1] == 'clients':\n"
            "    print(Path(os.environ['TEST_CLIENTS']).read_text())\n"
            "else:\n"
            "    with Path(os.environ['TEST_ACTIONS']).open('a') as out:\n"
            "        out.write(json.dumps(sys.argv[1:]) + '\\n')\n"
        )
        hyprctl.chmod(0o755)

    def notification(self):
        args = json.loads(self.capture.read_text())
        return args[:args.index("--exec")]

    def click_notification(self):
        args = json.loads(self.capture.read_text())
        return subprocess.run(args[args.index("--exec") + 1:], env=self.env,
                              capture_output=True, text=True)

    def install_test_windows(self):
        target = {"address": "0x123", "pid": os.getpid(), "stableId": "target"}
        other = {"address": "0x456", "pid": 99999999, "stableId": "other"}
        self.clients.write_text(json.dumps([other, target]))
        return target

    def run_script(self, relative, *args):
        return subprocess.run([sys.executable, str(ROOT / relative), *args],
                              env=self.env, capture_output=True, text=True)

    def test_completion_has_project_and_escaped_bounded_preview(self):
        payload = {"type": "agent-turn-complete", "cwd": "/tmp/my project",
                   "last-assistant-message": "Done <b>&\n" + "x" * 500}
        result = self.run_script("src/omarchy-codex-notify", json.dumps(payload))
        self.assertEqual(result.returncode, 0, result.stderr)
        args = self.notification()
        self.assertEqual(args[:2], ["notification", "send"])
        self.assertEqual(args[-2], "Codex terminou · my project")
        self.assertTrue(args[-1].startswith("Done &lt;b&gt;&amp; "))
        self.assertTrue(args[-1].endswith("…"))
        self.assertLess(len(args[-1]), 260)
        self.assertTrue(any(arg.startswith("--icon=") and "codex-desktop" in arg for arg in args))

    def test_internal_title_generation_does_not_notify(self):
        payload = {"type": "agent-turn-complete", "client": "codex-tui",
                   "input-messages": ["Generate a concise, single-line task title of at most 36 characters and under five words where possible.\n\nUser prompt:\nTest"],
                   "last-assistant-message": '{"title":"Validate notifications"}'}
        result = self.run_script("src/omarchy-codex-notify", json.dumps(payload))
        self.assertEqual(result.returncode, 0, result.stderr)
        self.assertFalse(self.capture.exists())

        payload["input-messages"] = ["Return JSON with a title field."]
        result = self.run_script("src/omarchy-codex-notify", json.dumps(payload))
        self.assertEqual(result.returncode, 0, result.stderr)
        self.assertIn("Validate notifications", self.notification()[-1])

    def test_other_events_are_ignored(self):
        result = self.run_script("src/omarchy-codex-notify", '{"type":"approval-requested"}')
        self.assertEqual(result.returncode, 0, result.stderr)
        self.assertFalse(self.capture.exists())

    def test_dash_leading_response_remains_notification_text(self):
        payload = {"type": "agent-turn-complete", "last-assistant-message": "--exec"}
        result = self.run_script("src/omarchy-codex-notify", json.dumps(payload))
        self.assertEqual(result.returncode, 0, result.stderr)
        self.assertEqual(self.notification()[-1], "\u200b--exec")

    def test_empty_completion_has_readable_fallback(self):
        result = self.run_script("src/omarchy-codex-notify", '{"type":"agent-turn-complete"}')
        self.assertEqual(result.returncode, 0, result.stderr)
        self.assertIn("Resposta pronta", self.notification()[-1])

    def test_click_focuses_originating_process_not_another_window(self):
        self.install_test_windows()
        result = self.run_script("src/omarchy-codex-notify", '{"type":"agent-turn-complete"}')
        self.assertEqual(result.returncode, 0, result.stderr)
        self.assertFalse(self.actions.exists(), "Delivery must not change focus")
        result = self.click_notification()
        self.assertEqual(result.returncode, 0, result.stderr)
        self.assertEqual(json.loads(self.actions.read_text()),
                         ["dispatch", 'hl.dsp.focus({ window = "address:0x123" })'])

    def test_click_ignores_closed_or_replaced_origin_window(self):
        target = self.install_test_windows()
        self.run_script("src/omarchy-codex-notify", '{"type":"agent-turn-complete"}')
        target["stableId"] = "replacement"
        self.clients.write_text(json.dumps([target]))
        self.assertEqual(self.click_notification().returncode, 0)
        self.assertFalse(self.actions.exists())

    def test_ambiguous_process_does_not_focus_an_arbitrary_window(self):
        target = self.install_test_windows()
        self.clients.write_text(json.dumps([target, dict(target, address="0x789")]))
        self.run_script("src/omarchy-codex-notify", '{"type":"agent-turn-complete"}')
        self.assertEqual(self.click_notification().returncode, 0)
        self.assertFalse(self.actions.exists())

    def test_tmux_click_restores_originating_session_window_and_pane(self):
        self.install_test_windows()
        self.env.update(TMUX="/tmp/test-tmux.sock,100,0", TMUX_PANE="%7")
        tmux = self.directory / "tmux"
        tmux.write_text(
            f"#!{sys.executable}\nimport json, os, sys\nfrom pathlib import Path\n"
            "command = sys.argv[3]\n"
            "if command == 'display-message':\n"
            "    print('4321\\t$2\\t@3')\n"
            "elif command == 'list-clients':\n"
            f"    print('{os.getpid()}\\t/dev/pts/42\\t$2\\t100')\n"
            "else:\n"
            "    with Path(os.environ['TEST_ACTIONS']).open('a') as out:\n"
            "        out.write(json.dumps(sys.argv[1:]) + '\\n')\n"
        )
        tmux.chmod(0o755)
        self.run_script("src/omarchy-codex-notify", '{"type":"agent-turn-complete"}')
        self.assertFalse(self.actions.exists())
        result = self.click_notification()
        self.assertEqual(result.returncode, 0, result.stderr)
        actions = [json.loads(line) for line in self.actions.read_text().splitlines()]
        self.assertEqual(actions, [
            ["-S", "/tmp/test-tmux.sock", "switch-client", "-c", "/dev/pts/42", "-t", "$2"],
            ["-S", "/tmp/test-tmux.sock", "select-window", "-t", "$2:@3"],
            ["-S", "/tmp/test-tmux.sock", "select-pane", "-t", "%7"],
            ["dispatch", 'hl.dsp.focus({ window = "address:0x123" })'],
        ])

    def test_invalid_payload_does_not_expose_contents(self):
        for payload in ("private invalid data", "[]", '{"type":"agent-turn-complete","cwd":42}'):
            with self.subTest(payload=payload):
                result = self.run_script("src/omarchy-codex-notify", payload)
                self.assertEqual(result.returncode, 2)
                self.assertNotIn(payload, result.stderr)
                self.assertFalse(self.capture.exists())

    def test_install_preserves_settings_backs_up_and_is_idempotent(self):
        config = self.directory / "config.toml"
        original = '# personal setting\nmodel = "test-model"\n[tui]\nnotifications = [\n "agent-turn-complete",\n "approval-requested",\n]\nstatus_line = ["git-branch"]\n'
        config.write_text(original)
        for _ in range(2):
            result = self.run_script("scripts/install-codex-notifications")
            self.assertEqual(result.returncode, 0, result.stderr)
        settings = tomllib.loads(config.read_text())
        self.assertEqual(settings["model"], "test-model")
        self.assertEqual(settings["tui"]["status_line"], ["git-branch"])
        self.assertEqual(settings["tui"]["notifications"], ["approval-requested"])
        self.assertEqual(settings["notify"], ["python3", str(ROOT / "src/omarchy-codex-notify")])
        backups = list(self.directory.glob("config.toml.bak.*"))
        self.assertEqual(len(backups), 1)
        self.assertEqual(backups[0].read_text(), original)

    def test_install_new_config_and_preserve_existing_notify(self):
        result = self.run_script("scripts/install-codex-notifications")
        self.assertEqual(result.returncode, 0, result.stderr)
        config = self.directory / "config.toml"
        original = 'notify = ["existing-notifier"]\n'
        config.write_text(original)
        result = self.run_script("scripts/install-codex-notifications")
        self.assertNotEqual(result.returncode, 0)
        self.assertEqual(config.read_text(), original)


if __name__ == "__main__":
    unittest.main()

File: __tests__/omarchy-codex-notify.integration.spec.ts
import assert from "node:assert/strict";
import { chmodSync, mkdtempSync, readFileSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { spawnSync } from "node:child_process";
import test from "node:test";
import type { TestContext } from "node:test";

const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), "..");

function executable(path: string, source: string): void {
  writeFileSync(path, `#!/usr/bin/env node\n${source}`);
  chmodSync(path, 0o755);
}

class Fixture {
  readonly directory = mkdtempSync(join(tmpdir(), "codex-notify-"));
  readonly capture = join(this.directory, "notification.json");
  readonly clients = join(this.directory, "clients.json");
  readonly actions = join(this.directory, "actions.jsonl");
  readonly env: Record<string, string>;

  constructor() {
    const inherited = Object.fromEntries(
      Object.entries(process.env).filter((entry): entry is [string, string] => entry[1] !== undefined),
    );
    this.env = {
      ...inherited,
      PATH: `${this.directory}:${inherited.PATH}`,
      NOTIFICATION_CAPTURE: this.capture,
      TEST_CLIENTS: this.clients,
      TEST_ACTIONS: this.actions,
    };
    delete this.env.TMUX;
    delete this.env.TMUX_PANE;
    writeFileSync(this.clients, "[]");
    executable(
      join(this.directory, "omarchy"),
      'const fs = require("node:fs");\nfs.writeFileSync(process.env.NOTIFICATION_CAPTURE, JSON.stringify(process.argv.slice(2)));\n',
    );
    executable(
      join(this.directory, "hyprctl"),
      'const fs = require("node:fs");\nif (process.argv[2] === "clients") {\n  process.stdout.write(fs.readFileSync(process.env.TEST_CLIENTS, "utf8"));\n} else {\n  fs.appendFileSync(process.env.TEST_ACTIONS, `${JSON.stringify(process.argv.slice(2))}\\n`);\n}\n',
    );
  }

  cleanup(): void {
    rmSync(this.directory, { recursive: true });
  }

  run(...args: string[]) {
    return spawnSync(process.execPath, [join(ROOT, "src/omarchy-codex-notify.ts"), ...args], {
      env: this.env,
      encoding: "utf8",
    });
  }

  notification(): string[] {
    const args: string[] = JSON.parse(readFileSync(this.capture, "utf8"));
    return args.slice(0, args.indexOf("--exec"));
  }

  clickNotification() {
    const args: string[] = JSON.parse(readFileSync(this.capture, "utf8"));
    const command = args.slice(args.indexOf("--exec") + 1);
    return spawnSync(command[0], command.slice(1), { env: this.env, encoding: "utf8" });
  }

  notificationTarget(): unknown {
    const args: string[] = JSON.parse(readFileSync(this.capture, "utf8"));
    return JSON.parse(args.at(-1) ?? "null");
  }

  installTestWindows(): Record<string, string | number> {
    const target = { address: "0x123", pid: process.pid, stableId: "target" };
    const other = { address: "0x456", pid: 99_999_999, stableId: "other" };
    writeFileSync(this.clients, JSON.stringify([other, target]));
    return target;
  }
}

function fixtureFor(context: TestContext): Fixture {
  const fixture = new Fixture();
  context.after(() => fixture.cleanup());
  return fixture;
}

test("completion has project and escaped bounded preview", (context) => {
  const fixture = fixtureFor(context);
  const payload = {
    type: "agent-turn-complete",
    cwd: "/tmp/my project",
    "last-assistant-message": `Done <b>&\n${"x".repeat(500)}`,
  };
  const result = fixture.run(JSON.stringify(payload));
  assert.equal(result.status, 0, result.stderr);
  const args = fixture.notification();
  assert.deepEqual(args.slice(0, 2), ["notification", "send"]);
  assert.equal(args.at(-2), "Codex terminou · my project");
  assert.match(args.at(-1) ?? "", /^Done &lt;b&gt;&amp; /);
  assert.match(args.at(-1) ?? "", /…$/);
  assert.ok((args.at(-1)?.length ?? Infinity) < 260);
  assert.ok(args.some((arg) => arg.startsWith("--icon=") && arg.includes("codex-desktop")));
});

test("internal title generation does not notify", (context) => {
  const fixture = fixtureFor(context);
  const payload = {
    type: "agent-turn-complete",
    client: "codex-tui",
    "input-messages": [
      "Generate a concise, single-line task title of at most 36 characters and under five words where possible.\n\nUser prompt:\nTest",
    ],
    "last-assistant-message": '{"title":"Validate notifications"}',
  };
  assert.equal(fixture.run(JSON.stringify(payload)).status, 0);
  assert.throws(() => readFileSync(fixture.capture));

  payload["input-messages"] = ["Return JSON with a title field."];
  assert.equal(fixture.run(JSON.stringify(payload)).status, 0);
  assert.match(fixture.notification().at(-1) ?? "", /Validate notifications/);
});

test("other events are ignored", (context) => {
  const fixture = fixtureFor(context);
  assert.equal(fixture.run('{"type":"approval-requested"}').status, 0);
  assert.throws(() => readFileSync(fixture.capture));
});

test("dash-leading response remains notification text", (context) => {
  const fixture = fixtureFor(context);
  const result = fixture.run('{"type":"agent-turn-complete","last-assistant-message":"--exec"}');
  assert.equal(result.status, 0, result.stderr);
  assert.equal(fixture.notification().at(-1), "\u200b--exec");
});

test("empty completion has readable fallback", (context) => {
  const fixture = fixtureFor(context);
  assert.equal(fixture.run('{"type":"agent-turn-complete"}').status, 0);
  assert.match(fixture.notification().at(-1) ?? "", /Resposta pronta/);
});

test("click focuses the originating process", (context) => {
  const fixture = fixtureFor(context);
  fixture.installTestWindows();
  assert.equal(fixture.run('{"type":"agent-turn-complete"}').status, 0);
  assert.deepEqual(fixture.notificationTarget(), {
    window: { address: "0x123", pid: process.pid, stableId: "target" },
  });
  assert.throws(() => readFileSync(fixture.actions));
  const click = fixture.clickNotification();
  assert.equal(click.status, 0, click.stderr);
  assert.deepEqual(JSON.parse(readFileSync(fixture.actions, "utf8")), [
    "dispatch",
    'hl.dsp.focus({ window = "address:0x123" })',
  ]);
});

test("click ignores a closed or replaced origin window", (context) => {
  const fixture = fixtureFor(context);
  const target = fixture.installTestWindows();
  fixture.run('{"type":"agent-turn-complete"}');
  target.stableId = "replacement";
  writeFileSync(fixture.clients, JSON.stringify([target]));
  assert.equal(fixture.clickNotification().status, 0);
  assert.throws(() => readFileSync(fixture.actions));
});

test("ambiguous process does not focus an arbitrary window", (context) => {
  const fixture = fixtureFor(context);
  const target = fixture.installTestWindows();
  writeFileSync(fixture.clients, JSON.stringify([target, { ...target, address: "0x789" }]));
  fixture.run('{"type":"agent-turn-complete"}');
  assert.equal(fixture.clickNotification().status, 0);
  assert.throws(() => readFileSync(fixture.actions));
});

test("tmux click restores the originating session, window, and pane", (context) => {
  const fixture = fixtureFor(context);
  fixture.installTestWindows();
  fixture.env.TMUX = "/tmp/test-tmux.sock,100,0";
  fixture.env.TMUX_PANE = "%7";
  executable(
    join(fixture.directory, "tmux"),
    `const fs = require("node:fs");\nconst command = process.argv[4];\nif (command === "display-message") {\n  console.log("4321\\t$2\\t@3");\n} else if (command === "list-clients") {\n  console.log("${process.pid}\\t/dev/pts/42\\t$2\\t100");\n} else {\n  fs.appendFileSync(process.env.TEST_ACTIONS, JSON.stringify(process.argv.slice(2)) + "\\n");\n}\n`,
  );
  fixture.run('{"type":"agent-turn-complete"}');
  assert.throws(() => readFileSync(fixture.actions));
  const click = fixture.clickNotification();
  assert.equal(click.status, 0, click.stderr);
  const actions = readFileSync(fixture.actions, "utf8").trim().split("\n").map((line) => JSON.parse(line));
  assert.deepEqual(actions, [
    ["-S", "/tmp/test-tmux.sock", "switch-client", "-c", "/dev/pts/42", "-t", "$2"],
    ["-S", "/tmp/test-tmux.sock", "select-window", "-t", "$2:@3"],
    ["-S", "/tmp/test-tmux.sock", "select-pane", "-t", "%7"],
    ["dispatch", 'hl.dsp.focus({ window = "address:0x123" })'],
  ]);
});

test("invalid payloads do not expose their contents", (context) => {
  const fixture = fixtureFor(context);
  for (const payload of ["private invalid data", "[]", '{"type":"agent-turn-complete","cwd":42}']) {
    const result = fixture.run(payload);
    assert.equal(result.status, 2);
    assert.doesNotMatch(result.stderr, new RegExp(payload.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")));
    assert.throws(() => readFileSync(fixture.capture));
  }
});
